### PR TITLE
[MC-1411] Zscaler v1.5.0

### PR DESCRIPTION
**Pull request by cdankowski-r7**
### Description

Describe the proposed changes:

  - Add actions: `Get Users`, `Delete User`,  `Create User`, `Get URL Category by Name` and `Update URLs of URL Category`

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: create_user\nCreating new user...\n\n",
    "meta": {},
    "output": {
      "user": {
        "comments": "Lorem ipsum",
        "deleted": false,
        "department": {
          "comments": "Lorem ipsum",
          "id": 12126432,
          "name": "Test Dep 1"
        },
        "email": "https://example.com",
        "groups": [
          {
            "comments": "Lorem ipsum",
            "id": 12126435,
            "name": "Test Group 1"
          }
        ],
        "id": 12143008,
        "name": "Test User 1",
        "tempAuthEmail": "https://example.com"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.5.0 --debug run < tests/create_user.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: delete_user\nDeleting user ID: 12156889.\n\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.5.0 --debug run < tests/delete_user.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: get_url_category_by_name\n",
    "meta": {},
    "output": {
      "urlCategory": {
        "customCategory": false,
        "customIpRangesCount": 0,
        "customUrlsCount": 4,
        "dbCategorizedUrls": [],
        "description": "BLOG_DESC",
        "editable": true,
        "id": "BLOG",
        "ipRangesRetainingParentCategoryCount": 0,
        "keywords": [],
        "keywordsRetainingParentCategory": [],
        "superCategory": "INTERNET_COMMUNICATION",
        "type": "URL_CATEGORY",
        "urls": [
          "https://example.com",
          "https://example.com",
          "https://example.com",
          "https://example.com"
        ],
        "urlsRetainingParentCategoryCount": 0,
        "val": 55
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.5.0 --debug run < tests/get_url_category_by_name.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: get_users\nGetting list of users with filter: {'department': 'Test Dep 1', 'group': 'Test Group 1', 'name': 'User', 'page': 1, 'pageSize': 50}.\n\n",
    "meta": {},
    "output": {
      "users": [
        {
          "adminUser": false,
          "comments": "Lorem ipsum",
          "deleted": false,
          "department": {
            "id": 12126432,
            "name": "Test Dep 1"
          },
          "email": "https://example.com",
          "groups": [
            {
              "id": 12126435,
              "name": "Test Group 1"
            }
          ],
          "id": 12143008,
          "isNonEditable": false,
          "name": "Test User 1",
          "tempAuthEmail": "https://example.com"
        },
        {
          "adminUser": false,
          "deleted": false,
          "department": {
            "id": 12126432,
            "name": "Test Dep 1"
          },
          "email": "https://example.com",
          "groups": [
            {
              "id": 12126435,
              "name": "Test Group 1"
            },
            {
              "id": 12126600,
              "name": "Test Group 2"
            }
          ],
          "id": 12143732,
          "isNonEditable": false,
          "name": "Test User Plugin 1"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.5.0 --debug run < tests/get_users.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: update_urls_of_url_category\n",
    "meta": {},
    "output": {
      "urlCategory": {
        "customCategory": false,
        "customIpRangesCount": 0,
        "customUrlsCount": 4,
        "dbCategorizedUrls": [],
        "description": "BLOG_DESC",
        "editable": true,
        "id": "BLOG",
        "ipRangesRetainingParentCategoryCount": 0,
        "keywords": [],
        "keywordsRetainingParentCategory": [],
        "type": "URL_CATEGORY",
        "urls": [
          "https://example.com",
          "https://example.com",
          "https://example.com",
          "https://example.com"
        ],
        "urlsRetainingParentCategoryCount": 0,
        "val": 55
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.5.0 --debug run < tests/update_urls_of_url_category.json
</summary>
</details>

### CODE REVIEW

**Review by nczapska-r7 — APPROVED** [review]

**Review by dsliwinski-r7 — APPROVED** [review]

File: plugins/zscaler/icon_zscaler/util/helpers.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+from insightconnect_plugin_runtime.exceptions import PluginException
+from re import sub, match, split
+from typing import Union
+from icon_zscaler.util.constants import Assistance, Cause
+
+_SNAKE_CASE_REGEX = r"\b[a-z0-9]+(_[a-z0-9]+)*\b"
+_CAMEL_CASE_REGEX = r"\b[a-z0-9]+([A-Z][a-z]+[0-9]*)*\b"
+_PASCAL_CASE_REGEX = r"\b[A-Z][a-z]+[0-9]*([A-Z][a-z]+[0-9]*)*\b"
+_CAMEL_CASE_ACRONYM_REGEX = r"\b[a-z0-9]+([A-Z]+[0-9]*)*\b"
```
Comment (igorski-r7) [inline plugins/zscaler/icon_zscaler/util/helpers.py:9]: I suggest removing the first underscore `_`
```suggestion
SNAKE_CASE_REGEX = r"\b[a-z0-9]+(_[a-z0-9]+)*\b"
CAMEL_CASE_REGEX = r"\b[a-z0-9]+([A-Z][a-z]+[0-9]*)*\b"
PASCAL_CASE_REGEX = r"\b[A-Z][a-z]+[0-9]*([A-Z][a-z]+[0-9]*)*\b"
CAMEL_CASE_ACRONYM_REGEX = r"\b[a-z0-9]+([A-Z]+[0-9]*)*\b"
```

File: plugins/zscaler/icon_zscaler/util/helpers.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def clean_dict(dictionary: dict) -> dict:
```
Comment (igorski-r7) [inline plugins/zscaler/icon_zscaler/util/helpers.py:12]: General typing is fine, but I suggest using more structured typing 😄 
`from typing import Any, Dict`
```suggestion
def clean_dict(dictionary: Dict[str, Any]) -> Dict[str, Any]:
```

File: plugins/zscaler/unit_test/expected/success.json.exp [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "success": true
+}
```
Comment (igorski-r7) [inline plugins/zscaler/unit_test/expected/success.json.exp:3]: FYI, remember to add new line at the end of files 😄

**Review by igorski-r7 — COMMENTED** [review]
LGTM! 🎉 
Left some small comments 😄

File: plugins/zscaler/icon_zscaler/util/helpers.py (hunk 1 above)
Comment (cdankowski-r7) [reply plugins/zscaler/icon_zscaler/util/helpers.py]: Updated. Thanks for your suggestion.

File: plugins/zscaler/icon_zscaler/util/helpers.py (hunk 2 above)
Comment (cdankowski-r7) [reply plugins/zscaler/icon_zscaler/util/helpers.py]: Updated. Thanks for your suggestion.

File: plugins/zscaler/unit_test/expected/success.json.exp (hunk 3 above)
Comment (cdankowski-r7) [reply plugins/zscaler/unit_test/expected/success.json.exp]: Updated. Thanks for your suggestion.

**Review by igorski-r7 — APPROVED** [review]

**Comment by igorski-r7** [comment]
Please make sure that all checks have passed 😄. 
In this case I suggest to change `hash` to for example `hash_` to maintain clarity 😄

**Pull request merged**